stats/statistics.go

```diff
@@ -0,0 +1,251 @@
+	Map() *expvar.Map
+	// A raw values map
+	Values() map[string]interface{}
+
+	// Set a level statistics to a particular integer value
+	SetInt(n string, i int64) Statistics
+	// Set a level statistic to a particular float value
+	SetFloat(n string, f float64) Statistics
+	// Set a string statistic
+	SetString(n string, s string) Statistics
+
+	// Add an int value to an int statistic
+	AddInt(n string, i int64) Statistics
+	// Add a float value to a float statistic
+	AddFloat(n string, f float64) Statistics
+
+	// The number of open references to the receiver.
+	Refs() int
+	// Drop one reference to the object obtained with Open(). T
+	// The described object closes the Statistics object first, then the monitor.
+	Close()
+}
+
+// The type which is used to implement both the Builder and Statistics interface
+type statistics struct {
+	mu         sync.RWMutex
+	registry   Registry
+	name       string
+	key        string
+	tags       map[string]string
+	impl       *expvar.Map
+	intVars    map[string]*expvar.Int
+	stringVars map[string]*expvar.String
+	floatVars  map[string]*expvar.Float
+	built      bool
+	refs       int
+}
+
+// Checks whether the receiver has already been built and returns an error if it has
+func (s *statistics) checkNotBuilt() error {
+	if s.built {
+		return ErrAlreadyBuilt
+	} else {
+		return nil
+	}
+}
+
+// Calls checkNotBuilt and panic if an error is returned
+func (s *statistics) assertNotBuilt() {
+	if err := s.checkNotBuilt(); err != nil {
+		panic(err)
+	}
+}
+
+// Declare an integer statistic
+func (s *statistics) DeclareInt(n string, iv int64) Builder {
+	s.assertNotBuilt()
+	v := &expvar.Int{}
+	v.Set(iv)
+	s.impl.Set(n, v)
+	s.intVars[n] = v
+	return s
+}
+
+// Declare a string statistic
+func (s *statistics) DeclareString(n string, iv string) Builder {
+	s.assertNotBuilt()
+	v := &expvar.String{}
+	v.Set(iv)
+	s.impl.Set(n, v)
+	s.stringVars[n] = v
+	return s
+}
+
+// Declare a float statistic
+func (s *statistics) DeclareFloat(n string, iv float64) Builder {
+	s.assertNotBuilt()
+	v := &expvar.Float{}
+	v.Set(iv)
+	s.impl.Set(n, v)
+	s.floatVars[n] = v
+	return s
+}
+
+// Finish building a Statistics returning an error on failure
+func (s *statistics) Build() (Openable, error) {
+	if err := s.checkNotBuilt(); err != nil {
+		return nil, err
+	}
+
+	s.built = true
+	tmp := &expvar.Map{}
+	tmp.Init()
+	s.impl.Do(func(kv expvar.KeyValue) {
+		tmp.Set(kv.Key, kv.Value)
+	})
+	s.impl = tmp
+
+	return s, nil
+}
+
+// Finish building a Statistics and panic on failure.
+func (s *statistics) MustBuild() Openable {
+	if set, err := s.Build(); err != nil {
+		panic(err)
+	} else {
+		return set
+	}
+}
+
+func (s *statistics) Key() string {
+	return s.key
+}
+
+func (s *statistics) Name() string {
+	return s.name
+}
+
+func (s *statistics) Tags() map[string]string {
+	return s.tags
+}
+
+func (s *statistics) Map() *expvar.Map {
+	return s.impl
+}
+
+func (s *statistics) Values() map[string]interface{} {
```

Comment:
Need to test what happens if someone has put a map into a map obtained by Statistics.Map(). Not a useful scenario, but behaviour should be well defined.